Final Project.


In [1]:
import numpy as np
import pandas as pd
import sklearn



In [129]:
# ! pip install nltk
# ! pip install TextBlob

     |████████████████████████████████| 636 kB 8.2 MB/s eta 0:00:01


In [2]:
import nltk
from textblob import TextBlob

In [3]:
avail = pd.read_csv("availabilities.csv")
host = pd.read_csv("hosts.csv")
# lst = pd.read_csv("listings.csv")
rev = pd.read_csv("reviews.csv")

In [4]:
# Had to strip the first line from the file - just a title line, then worked fine.  *****
lst = pd.read_csv('listings.csv')

In [107]:
avail

listing_id      date available     price  minimum_nights  \
0              2384  11/21/19         t   $65.00                2   
1              2384  11/22/19         f   $65.00                2   
2              2384  11/23/19         f   $65.00                2   
3              2384  11/24/19         f   $65.00                2   
4              2384  11/25/19         f   $65.00                2   
...             ...       ...       ...       ...             ...   
1048570    21788299    7/8/20         f  $239.00                2   
1048571    21788299    7/9/20         f  $239.00                2   
1048572    21788299   7/10/20         f  $239.00                2   
1048573    21788299   7/11/20         f  $239.00                2   
1048574    21788299   7/12/20         f  $239.00                2   

         maximum_nights  
0                  1125  
1                  1125  
2                  1125  
3                  1125  
4                  1125  
...                 ...  
1048570            1125  
1048571            1125  
1048572            1125  
1048573            1125  
1048574            1125  

[1048575 rows x 6 columns]

In [108]:
avail.isnull()
avail.isna().sum()

listing_id        0
date              0
available         0
price             0
minimum_nights    0
maximum_nights    0
dtype: int64

In [109]:
host

id                                     host_url  \
0          2613       https://www.airbnb.com/users/show/2613   
1          5775       https://www.airbnb.com/users/show/5775   
2         17928      https://www.airbnb.com/users/show/17928   
3         33004      https://www.airbnb.com/users/show/33004   
4          2140       https://www.airbnb.com/users/show/2140   
...         ...                                          ...   
4622  285121807  https://www.airbnb.com/users/show/285121807   
4623   50982462   https://www.airbnb.com/users/show/50982462   
4624  114503989  https://www.airbnb.com/users/show/114503989   
4625    2912333    https://www.airbnb.com/users/show/2912333   
4626   50413447   https://www.airbnb.com/users/show/50413447   

                  name host_since                          location  \
0              Rebecca    8/29/08  Chicago, Illinois, United States   
1     Craig & Kathleen   12/29/08  Chicago, Illinois, United States   
2                Sarah    5/19/09  Chicago, Illinois, United States   
3          At Home Inn    8/21/09  Chicago, Illinois, United States   
4                 Lois    8/16/08   Aurora, Illinois, United States   
...                ...        ...                               ...   
4622             Terri    8/12/19  Chicago, Illinois, United States   
4623             Brian   12/10/15  Chicago, Illinois, United States   
4624            Jesson     2/2/17                             China   
4625          Benjamin    7/12/12  Chicago, Illinois, United States   
4626         Christian    12/2/15  Chicago, Illinois, United States   

           response_time superhost  
0         within an hour         t  
1     within a few hours         t  
2         within an hour         t  
3         within an hour         f  
4         within an hour         f  
...                  ...       ...  
4622                 NaN         f  
4623                 NaN         f  
4624                 NaN         f  
4625                 NaN         f  
4626      within an hour         t  

[4627 rows x 7 columns]

In [110]:
host.isna().sum()

id                 0
host_url           0
name               0
host_since         0
location          10
response_time    767
superhost          0
dtype: int64

In [5]:
lst

id                            listing_url  \
0         2384      https://www.airbnb.com/rooms/2384   
1         4505      https://www.airbnb.com/rooms/4505   
2         7126      https://www.airbnb.com/rooms/7126   
3         9811      https://www.airbnb.com/rooms/9811   
4        10610     https://www.airbnb.com/rooms/10610   
...        ...                                    ...   
8528  40262251  https://www.airbnb.com/rooms/40262251   
8529  40263536  https://www.airbnb.com/rooms/40263536   
8530  40264107  https://www.airbnb.com/rooms/40264107   
8531  40264148  https://www.airbnb.com/rooms/40264148   
8532  40264411  https://www.airbnb.com/rooms/40264411   

                                                   name  \
0     Walk to UChicago, 10 min train to McCormick Place   
1     One Great Apartment. 384 Great Reviews. 1 Bad ...   
2                   Tiny Studio Apartment 94 Walk Score   
3                         Barbara's Hideaway - Old Town   
4                      3 Comforts of Cooperative Living   
...                                                 ...   
8528  Perfect Old Town condo you will love. Come stay!!   
8529  ☼O'Hare-1Bed-Private Bath-Rosemont-AllState Arena   
8530    One queen bed room in cozy house northwest side   
8531  ☼2bed 1bath O'hare, Allstate arena,Casino,Rose...   
8532      Modern Private Room/ quick walk to the train!   

                                            description    host_id host_since  \
0     Visiting The University of Chicago or attendin...       2613    8/29/08   
1     We took the winter off, made things nicer arou...       5775   12/29/08   
2     A very small studio in a wonderful neighborhoo...      17928    5/19/09   
3     One-bedroom hideaway tucked into Old Town step...      33004    8/21/09   
4     The condo is the 2nd floor in a lovely 1912 3-...       2140    8/16/08   
...                                                 ...        ...        ...   
8528  This is a beautiful million-dollar condo, perf...    2912333    7/12/12   
8529  Is this your first booking on Airbnb? Use this...   50413447    12/2/15   
8530  Nice and quiet neighborhood, save a bunch agai...   35065108     6/5/15   
8531  Is this your first booking on Airbnb? Use this...   50413447    12/2/15   
8532  Enjoy a clean & comfortable room in a 1bd 2.5b...  178275885    3/13/18   

          neighbourhood         city state zipcode  ... cleaning_fee  \
0             Hyde Park      Chicago    IL   60637  ...        $0.00   
1        Little Village      Chicago    IL   60608  ...       $25.00   
2     Ukrainian Village      Chicago    IL   60622  ...        $0.00   
3              Old Town      Chicago    IL   60614  ...       $95.00   
4             Hyde Park      Chicago    IL   60615  ...       $50.00   
...                 ...          ...   ...     ...  ...          ...   
8528                NaN      Chicago    IL   60610  ...          NaN   
8529                NaN  Des Plaines    IL   60018  ...          NaN   
8530                NaN      Chicago    IL   60634  ...       $12.00   
8531                NaN  Des Plaines    IL   60018  ...        $0.00   
8532                NaN      Chicago    IL   60608  ...       $10.00   

      review_scores_rating  review_scores_accuracy review_scores_cleanliness  \
0                    100.0                    10.0                      10.0   
1                     96.0                    10.0                       9.0   
2                     93.0                    10.0                       9.0   
3                     92.0                    10.0                       9.0   
4                     82.0                     8.0                       7.0   
...                    ...                     ...                       ...   
8528                   NaN                     NaN                       NaN   
8529                   NaN                     NaN                       NaN   
8530                   NaN                     NaN          

In [11]:
type(lst)

pandas.core.frame.DataFrame

In [6]:
lst.isna().sum()

id                                0
listing_url                       0
name                              0
description                     129
host_id                           0
host_since                        0
neighbourhood                    27
city                              3
state                            23
zipcode                          22
country_code                      0
latitude                          0
longitude                         0
property_type                     0
room_type                         0
accommodates                      0
bathrooms                         3
bedrooms                          4
security_deposit               2086
cleaning_fee                    836
review_scores_rating           1244
review_scores_accuracy         1244
review_scores_cleanliness      1244
review_scores_checkin          1245
review_scores_communication    1246
review_scores_location         1245
review_scores_value            1245
instant_bookable            

In [113]:
lst.dtypes

id                               int64
listing_url                     object
name                            object
description                     object
host_id                          int64
host_since                      object
neighbourhood                   object
city                            object
state                           object
zipcode                         object
country_code                    object
latitude                       float64
longitude                      float64
property_type                   object
room_type                       object
accommodates                     int64
bathrooms                      float64
bedrooms                       float64
security_deposit                object
cleaning_fee                    object
review_scores_rating           float64
review_scores_accuracy         float64
review_scores_cleanliness      float64
review_scores_checkin          float64
review_scores_communication    float64
review_scores_location   

In [114]:
rev.isna().sum()

id                 0
listing_id         0
date               0
reviewer_id        0
reviewer_name      0
comments         251
dtype: int64

In [115]:
rev.iloc[233904]

id                                             219458638
listing_id                                      21784716
date                                            12/17/17
reviewer_id                                    157831350
reviewer_name                                    Mileena
comments         The host, area and the unit is perfect!
Name: 233904, dtype: object

In [7]:
# SLICE out the records for PJO. *********

revP = rev.iloc[116840:233904]

revP


id  listing_id      date  reviewer_id reviewer_name  \
116840   69989174    12170773   4/15/16     65467524     Elizabeth   
116841   70277659    12170773   4/17/16     56990266       Pranith   
116842   70615667    12170773   4/19/16     66514184         Katie   
116843   71131787    12170773   4/23/16     46343079         Alana   
116844   71376925    12170773   4/25/16     61131477          Beth   
...           ...         ...       ...          ...           ...   
233899  556868477    21783531  10/30/19      8520589       Lindsey   
233900  559780590    21783531   11/4/19     72254229        Howard   
233901  566148318    21783531  11/18/19     42144708        Rafael   
233902  215412883    21784716  11/29/17     13109574        Justin   
233903  217997495    21784716  12/10/17     46340737           Zac   

                                                 comments  
116840  The condo was amazing. It was very spacious an...  
116841  Jessica was available on texts most of the tim...  
116842  Jessica is great, fast response! Our place was...  
116843  My family and I loved this apartment. It was v...  
116844  We have a lovely weekend in this rental! It wa...  
...                                                   ...  
233899  Super clean listing that's very accessible to ...  
233900  Great Location! We will definitely book this u...  
233901  Great location, quite space, has a king size b...  
233902  (Website hidden by Airbnb) where do I begin? L...  
233903  Awesome place, great neighborhood. It's right ...  

[117064 rows x 6 columns]

In [8]:
revP.isna().sum()

id                0
listing_id        0
date              0
reviewer_id       0
reviewer_name     0
comments         74
dtype: int64

In [9]:
revP.sample(10)

id  listing_id      date  reviewer_id reviewer_name  \
168406  413984689    16054326   2/18/19     59786612         Blake   
165593  199718439    15771760   10/2/17     10282622       Gregory   
205657  190249282    19827093    9/3/17    145784216         Bryce   
220844  299760707    20818822   7/31/18     33829467      Philippe   
154524  148910567    15060526    5/1/17     16342571        Merone   
144759  182857274    14238448   8/15/17      4758611       Maureen   
233769  486515063    21783333   7/12/19    117177653        Alyssa   
136018  336767391    13656318  10/14/18      1822518    Meghan Moe   
185827  412500104    18040546   2/15/19    227829522      Jeronimo   
166254  488099557    15871357   7/14/19      5257371       Natalia   

                                                 comments  
168406  Comfortable and cozy. Close to the train stati...  
165593  This has possibly been our best Airbnb stay ev...  
205657  The hosts were very hospitable. They allowed m...  
220844  Appartement grand et confortable, Metro et com...  
154524                              Very clean and nice.   
144759  Daniel's place fit what we needed at the right...  
233769  Ian's home was AMAZING. The modern touches, lu...  
136018                                   Fantastic space!  
185827  Great place very stylish and amazing shower an...  
166254  Great place. Super quiet, super comfy, super s...

In [35]:
Installed to determine the text of the comment.

# ! pip install langdetect

     |████████████████████████████████| 981 kB 7.0 MB/s eta 0:00:01
Using legacy 'setup.py install' for langdetect, since package 'wheel' is not installed.
    Running setup.py install for langdetect ... done


In [123]:
# from langdetect import detect

In [142]:
# try it

detect(revP['comments'].iloc[116952])

# issues with langdetect  

'en'

In [137]:
# TRY textblob

# b = TextBlob(revP['comments'].iloc[10])
# xx= b.detect_language()
# xx

'en'

In [134]:
# Get rid of the NaNs in Comments - replace with No Comment Added?  MAYBE DROP LATER

# revP['comments'] = revP['comments'].replace(np.nan, 'No Comment Added')

In [191]:
revP.iloc[3500]

id                                                     277094047.0
listing_id                                              12489153.0
date                                                       6/15/18
reviewer_id                                             56390780.0
reviewer_name                                                  Jan
comments         This home is in the perfect location for me to...
Name: 120340, dtype: object

In [195]:
# Records to be cleaned!  1281, 3328, 3329, 3499

# revP.loc[1281, 'comments'] = 'Smile emoji added'

# revP['comments'].iloc[1281] = 'Smile emoji added'
# revP['comments'].iloc[3328] = 'User rated the property as 4.5 out of 5.'
# revP['comments'].iloc[3329] = 'User simply entered a dash'


# len(revP['comments'].iloc[3329])

26

In [237]:
# Loop through, find any NON-English  TEST

# list of records NOT English
not_eng = []
maybe_bad = []

for i in range(0,117064):
    
#     if revP['comments'].iloc[i] != np.nan:
#         # b = TextBlob(revP['comments'].iloc[i])
#         # lang = b.detect_language()
#         lang = detect(revP['comments'].iloc[i])
#         print(i, lang)
   
# #         if lang != 'en':
# #             print(lang, i)
            
#     if revP['comments'].iloc[i].isnull(): # == np.nan:
#         print('NaN',i)

    if pd.notnull(revP['comments'].iloc[i]):
        

        com_string = revP['comments'].iloc[i]
    
        if com_string[0].isalpha():
            lang = detect(revP['comments'].iloc[i])
                # print(i, lang)
            if lang != 'en':
                    # print(lang, i)
                not_eng.append([i, lang])
        else:
            maybe_bad.append(i)
      

In [236]:
com_string = revP['comments'].iloc[4500]
com_string[0].isalpha()


True

In [239]:
# not_eng[-1]
# maybe_bad

len(not_eng)
# len(maybe_bad)

4680

In [240]:
maybe_bad

[224,
 429,
 465,
 547,
 1237,
 1413,
 1540,
 1786,
 2226,
 2712,
 2852,
 3013,
 3499,
 3668,
 3807,
 3959,
 4024,
 4097,
 4194,
 4362,
 4556,
 4625,
 4696,
 4735,
 4930,
 5043,
 5218,
 5654,
 5668,
 5819,
 5922,
 6637,
 6704,
 6766,
 6785,
 7174,
 7175,
 7516,
 7552,
 7714,
 9062,
 9103,
 9283,
 9458,
 9964,
 10194,
 10534,
 10586,
 10712,
 11303,
 11671,
 11924,
 11991,
 12416,
 12519,
 12624,
 12741,
 12774,
 12828,
 13113,
 13135,
 13485,
 13541,
 13559,
 13579,
 13640,
 14289,
 14354,
 14574,
 14929,
 15116,
 15237,
 15911,
 15944,
 16041,
 16129,
 16135,
 16205,
 16244,
 16514,
 16753,
 16820,
 16876,
 16965,
 17035,
 17128,
 17159,
 17266,
 17445,
 17945,
 18192,
 18347,
 19023,
 19253,
 19318,
 19548,
 19735,
 19827,
 19840,
 19935,
 19943,
 19980,
 21281,
 21408,
 21492,
 21503,
 21590,
 21860,
 22095,
 22153,
 22223,
 22520,
 22954,
 23149,
 23173,
 24013,
 24077,
 24367,
 25244,
 25267,
 25335,
 25628,
 26247,
 26802,
 26989,
 27083,
 27334,
 27836,
 28048,
 28112,
 28144,
 

In [244]:
not_eng

[[42, 'de'],
 [56, 'ro'],
 [64, 'ro'],
 [69, 'tl'],
 [135, 'ro'],
 [169, 'fr'],
 [177, 'ro'],
 [231, 'fr'],
 [234, 'ro'],
 [326, 'zh-cn'],
 [397, 'fr'],
 [401, 'de'],
 [404, 'pt'],
 [418, 'fr'],
 [458, 'de'],
 [463, 'es'],
 [464, 'fr'],
 [519, 'ko'],
 [537, 'es'],
 [542, 'ko'],
 [567, 'fr'],
 [568, 'es'],
 [570, 'es'],
 [579, 'es'],
 [607, 'ca'],
 [613, 'es'],
 [626, 'es'],
 [679, 'ca'],
 [681, 'ko'],
 [684, 'fi'],
 [734, 'fr'],
 [851, 'de'],
 [860, 'es'],
 [878, 'pt'],
 [882, 'nl'],
 [885, 'es'],
 [888, 'fr'],
 [905, 'ca'],
 [927, 'es'],
 [957, 'fr'],
 [997, 'fr'],
 [1052, 'it'],
 [1075, 'zh-cn'],
 [1081, 'zh-cn'],
 [1088, 'ro'],
 [1119, 'es'],
 [1123, 'es'],
 [1184, 'es'],
 [1185, 'de'],
 [1191, 'fr'],
 [1193, 'zh-cn'],
 [1199, 'es'],
 [1206, 'fr'],
 [1207, 'es'],
 [1220, 'es'],
 [1232, 'ja'],
 [1249, 'es'],
 [1251, 'es'],
 [1255, 'zh-cn'],
 [1281, 'sl'],
 [1336, 'es'],
 [1338, 'zh-cn'],
 [1347, 'zh-cn'],
 [1350, 'es'],
 [1396, 'zh-cn'],
 [1412, 'it'],
 [1454, 'cs'],
 [1480, 'cy'],
 

In [241]:
revP.iloc[224]

id                                                     337548362.0
listing_id                                              12170773.0
date                                                      10/16/18
reviewer_id                                              9891737.0
reviewer_name                                             Brittany
comments         5 stars all the way! Unique space, but still f...
Name: 117064, dtype: object

In [270]:
# Try to use googletrans

# ! pip install googletrans

# ! pip install git+https://github.com/BoseCorp/py-googletrans.git --upgrade

# ! pip install googletrans==3.1.0a0

# ! pip install pygoogletranslation

     |████████████████████████████████| 241 kB 9.5 MB/s eta 0:00:01
     |████████████████████████████████| 77 kB 10.4 MB/s eta 0:00:01
Using legacy 'setup.py install' for docx2txt, since package 'wheel' is not installed.
Using legacy 'setup.py install' for PyPDF2, since package 'wheel' is not installed.
    Running setup.py install for PyPDF2 ... done
    Running setup.py install for docx2txt ... done


In [272]:
# This version of Google translate allowed for chunks of reviews to be translated - approx. 1200 at a time - 
# before erroring out.  Four iterations on my part translated the entire file.

from pygoogletranslation import Translator

In [279]:
# Test Code
# _________

# translator = Translator()

# xx = revP['comments'].iloc[42]
# # xx = 'Die Unterkunft war sehr schön, Sauber und relativ zentral gelegen,'
# print(xx)
# xx = xx.replace(".", ",")
# translated = translator.translate(xx)

# print(translated.text)

# # translator.translate('Je suis un peu', dest='fr')

Die Unterkunft war sehr schön. Sauber und relativ zentral gelegen.
The accommodation was very nice, clean and relatively centrally located,


In [334]:
not_eng

[[42, 'de'],
 [56, 'ro'],
 [64, 'ro'],
 [69, 'tl'],
 [135, 'ro'],
 [169, 'fr'],
 [177, 'ro'],
 [231, 'fr'],
 [234, 'ro'],
 [326, 'zh-cn'],
 [397, 'fr'],
 [401, 'de'],
 [404, 'pt'],
 [418, 'fr'],
 [458, 'de'],
 [463, 'es'],
 [464, 'fr'],
 [519, 'ko'],
 [537, 'es'],
 [542, 'ko'],
 [567, 'fr'],
 [568, 'es'],
 [570, 'es'],
 [579, 'es'],
 [607, 'ca'],
 [613, 'es'],
 [626, 'es'],
 [679, 'ca'],
 [681, 'ko'],
 [684, 'fi'],
 [734, 'fr'],
 [851, 'de'],
 [860, 'es'],
 [878, 'pt'],
 [882, 'nl'],
 [885, 'es'],
 [888, 'fr'],
 [905, 'ca'],
 [927, 'es'],
 [957, 'fr'],
 [997, 'fr'],
 [1052, 'it'],
 [1075, 'zh-cn'],
 [1081, 'zh-cn'],
 [1088, 'ro'],
 [1119, 'es'],
 [1123, 'es'],
 [1184, 'es'],
 [1185, 'de'],
 [1191, 'fr'],
 [1193, 'zh-cn'],
 [1199, 'es'],
 [1206, 'fr'],
 [1207, 'es'],
 [1220, 'es'],
 [1232, 'ja'],
 [1249, 'es'],
 [1251, 'es'],
 [1255, 'zh-cn'],
 [1281, 'sl'],
 [1336, 'es'],
 [1338, 'zh-cn'],
 [1347, 'zh-cn'],
 [1350, 'es'],
 [1396, 'zh-cn'],
 [1412, 'it'],
 [1454, 'cs'],
 [1480, 'cy'],
 

In [320]:
# TEST Code
# ---------

# cmt = revP['comments'].iloc[1940]
# cmt = cmt.replace(".", ",")
    
# tr = translator.translate(cmt)
    
# revP['comments'].iloc[1940] = tr.text
# # tr.text
# # revP.loc[revP['comments'].iloc[1940]] = tr.text

In [11]:
revP['comments'].iloc[1940]

'房东是很和善的一对夫妻｡他们有很好的沟通能力,在我启程之前,他们用了一些时间来指导我顺利到达他们的家｡他们的房子是一栋三层小楼,房间很高,干净明亮｡这里有客人专用的卫生间､厨房､冰箱和客厅,很方便客人自己使用｡免费提供咖啡和报纸｡楼下有小超市,中医保健药店,中国美食餐馆｡公交车站就在门口,坐上公交车一会儿就到芝加哥市中心,不远处还有地铁,差不多一小时就可以到奥黑尔机场,出行很方便｡房东人很好相处,我非常感谢房东这几日的关照｡'

In [322]:
# try to get it all translated and updated.  PASS #1
# ---------------------------------------------------

# for i in not_eng:
    
#     rec_num = i[0]
    
    
#     cmt = revP['comments'].iloc[rec_num]
#     cmt = cmt.replace(".", ",")
    
#     tr = translator.translate(cmt)
    
#     revP['comments'].iloc[rec_num] = tr.text
    


Exception: Unexpected status code 429 from https://translate.google.com/_/TranslateWebserverUi/data/batchexecute after retried 3 loop with 5s delay

In [341]:
# The translate timed out around 950 records in.  Slice the list and try again.
# ----------------------------------------------------------------------------

# not_eng2 = not_eng[999:]
# not_eng2

[[24371, 'es'],
 [24444, 'es'],
 [24455, 'da'],
 [24478, 'ro'],
 [24545, 'es'],
 [24559, 'fr'],
 [24596, 'de'],
 [24633, 'af'],
 [24647, 'de'],
 [24652, 'fr'],
 [24677, 'de'],
 [24691, 'sv'],
 [24710, 'ko'],
 [24732, 'fr'],
 [24740, 'ro'],
 [24752, 'es'],
 [24790, 'it'],
 [24836, 'ro'],
 [24858, 'fr'],
 [24913, 'ro'],
 [24932, 'es'],
 [24942, 'ro'],
 [24980, 'ro'],
 [25030, 'fr'],
 [25064, 'es'],
 [25102, 'fr'],
 [25154, 'es'],
 [25169, 'de'],
 [25190, 'pt'],
 [25195, 'fr'],
 [25208, 'cy'],
 [25239, 'es'],
 [25240, 'no'],
 [25254, 'ro'],
 [25288, 'ko'],
 [25322, 'af'],
 [25355, 'fr'],
 [25390, 'es'],
 [25417, 'af'],
 [25428, 'ro'],
 [25436, 'es'],
 [25527, 'zh-cn'],
 [25537, 'de'],
 [25674, 'fr'],
 [25705, 'fr'],
 [25710, 'fr'],
 [25714, 'fr'],
 [25724, 'ca'],
 [25736, 'zh-cn'],
 [25751, 'zh-cn'],
 [25758, 'de'],
 [25763, 'ja'],
 [25799, 'af'],
 [25801, 'zh-cn'],
 [25812, 'es'],
 [25823, 'es'],
 [25831, 'fr'],
 [25841, 'es'],
 [25845, 'zh-cn'],
 [25847, 'es'],
 [25854, 'fr'],
 [25855, 

In [344]:
# try to get it all translated and updated.   PASS #2
# ---------------------------------------------------

# for i in not_eng2:
    
#     rec_num = i[0]
    
#     cmt = revP['comments'].iloc[rec_num]
#     cmt = cmt.replace(".", ",")
    
#     tr = translator.translate(cmt)
    
#     revP['comments'].iloc[rec_num] = tr.text
    

Exception: Unexpected status code 429 from https://translate.google.com/_/TranslateWebserverUi/data/batchexecute after retried 3 loop with 5s delay

In [375]:
# Slice to Prep for pass #3
# -------------------------

# not_eng3 = not_eng2[1240:]
# len(not_eng3)

2441

In [12]:
revP['comments'].iloc[55918]

'房东大姐人非常nice,很贴心,房子干净､整洁､舒适,位置也非常好,靠近城铁,路边停车也很方便,附近有便利店,推荐!'

In [378]:
# try to get it all translated and updated  PASS # 3
# --------------------------------------------------

# for i in not_eng3:
    
#     rec_num = i[0]
    
#     cmt = revP['comments'].iloc[rec_num]
#     cmt = cmt.replace(".", ",")
    
#     tr = translator.translate(cmt)
    
#     revP['comments'].iloc[rec_num] = tr.text
#     print(i)
    

[55913, 'fr']
[55917, 'fr']
[55918, 'zh-cn']
[55923, 'ko']
[55931, 'de']
[55935, 'zh-cn']
[55937, 'zh-cn']
[55959, 'zh-cn']
[55962, 'zh-cn']
[55976, 'fr']
[55995, 'no']
[55996, 'zh-cn']
[56013, 'ko']
[56028, 'ko']
[56029, 'zh-cn']
[56044, 'zh-cn']
[56047, 'af']
[56050, 'zh-cn']
[56060, 'de']
[56062, 'zh-cn']
[56066, 'hr']
[56068, 'ro']
[56074, 'fr']
[56080, 'ko']
[56092, 'fr']
[56117, 'af']
[56130, 'de']
[56205, 'fr']
[56279, 'es']
[56288, 'fr']
[56377, 'fr']
[56395, 'fr']
[56404, 'fr']
[56420, 'fr']
[56422, 'es']
[56450, 'ko']
[56452, 'ko']
[56468, 'pl']
[56550, 'es']
[56571, 'es']
[56652, 'fr']
[56690, 'pt']
[56724, 'fr']
[56794, 'ru']
[56797, 'es']
[56807, 'nl']
[56815, 'fr']
[56862, 'es']
[56870, 'fr']
[56875, 'es']
[56895, 'zh-cn']
[56896, 'es']
[56910, 'es']
[56919, 'es']
[56970, 'ko']
[57030, 'it']
[57102, 'af']
[57110, 'af']
[57123, 'fr']
[57221, 'es']
[57224, 'nl']
[57234, 'es']
[57236, 'cy']
[57239, 'de']
[57253, 'es']
[57254, 'de']
[57257, 'de']
[57265, 'ko']
[57275, 'es']
[

[70593, 'es']
[70610, 'es']
[70706, 'tl']
[70708, 'de']
[70716, 'es']
[70719, 'es']
[70730, 'fr']
[70765, 'cs']
[70771, 'cs']
[70778, 'fr']
[70819, 'hr']
[70823, 'es']
[70878, 'es']
[70888, 'es']
[70913, 'af']
[70914, 'af']
[70963, 'pt']
[70964, 'ro']
[70967, 'es']
[70973, 'ca']
[70978, 'fr']
[70986, 'ro']
[71004, 'it']
[71034, 'hr']
[71049, 'de']
[71052, 'es']
[71058, 'it']
[71089, 'ro']
[71134, 'es']
[71279, 'ko']
[71286, 'ro']
[71291, 'cy']
[71312, 'fr']
[71408, 'pt']
[71445, 'fr']
[71506, 'ko']
[71543, 'es']
[71548, 'ro']
[71555, 'ro']
[71559, 'es']
[71625, 'es']
[71630, 'es']
[71635, 'es']
[71641, 'so']
[71731, 'ro']
[71749, 'de']
[71761, 'ko']
[71809, 'ro']
[71862, 'no']
[71869, 'es']
[71902, 'pl']
[71925, 'es']
[71990, 'no']
[72071, 'af']
[72092, 'ro']
[72097, 'af']
[72109, 'es']
[72124, 'es']
[72135, 'es']
[72183, 'ro']
[72193, 'es']
[72265, 'tl']
[72285, 'ko']
[72292, 'ko']
[72300, 'es']
[72446, 'it']
[72534, 'de']
[72541, 'fr']
[72554, 'es']
[72555, 'fr']
[72611, 'ro']
[72614

[86044, 'de']
[86046, 'es']
[86051, 'es']
[86057, 'es']
[86064, 'es']
[86078, 'zh-cn']
[86164, 'zh-cn']
[86229, 'de']
[86245, 'ro']
[86270, 'de']
[86277, 'de']
[86294, 'es']
[86325, 'fr']
[86358, 'ro']
[86359, 'de']
[86367, 'fr']
[86370, 'fr']
[86383, 'pt']
[86395, 'es']
[86429, 'es']
[86446, 'fr']
[86454, 'fr']
[86464, 'es']
[86470, 'es']
[86477, 'ko']
[86484, 'es']
[86485, 'it']
[86487, 'tl']
[86492, 'cs']
[86563, 'es']
[86568, 'fr']
[86573, 'es']
[86574, 'ru']
[86627, 'ro']
[86654, 'fr']
[86729, 'es']
[86749, 'es']
[86762, 'es']
[86845, 'es']
[86921, 'ro']
[86959, 'de']
[87014, 'ro']
[87107, 'fr']
[87135, 'zh-cn']
[87153, 'es']
[87193, 'fr']
[87209, 'so']
[87223, 'ro']
[87261, 'es']
[87282, 'fr']
[87319, 'es']
[87382, 'de']
[87426, 'no']
[87441, 'af']
[87450, 'es']
[87467, 'es']
[87493, 'pl']
[87503, 'es']
[87521, 'es']
[87524, 'es']
[87541, 'af']
[87545, 'es']
[87548, 'so']
[87557, 'es']
[87564, 'ro']
[87581, 'sw']
[87647, 'ro']
[87664, 'ru']
[87669, 'ru']
[87671, 'fr']
[87697, 'de

Exception: Unexpected status code 429 from https://translate.google.com/_/TranslateWebserverUi/data/batchexecute after retried 3 loop with 5s delay

In [379]:
len(not_eng3)

2441

In [380]:
x = 0
for i in not_eng3:
    x += 1
    if i[0] == 88345:
        print(x)
    

1261


In [382]:
not_eng3[1260]

[88345, 'ro']

In [384]:
# That batch did 1260 records.  Onto Batch #4
# -------------------------------------------

# not_eng4 = not_eng3[1260:]

In [385]:
not_eng4

[[88345, 'ro'],
 [88381, 'de'],
 [88383, 'tl'],
 [88386, 'so'],
 [88387, 'ro'],
 [88400, 'de'],
 [88445, 'cs'],
 [88494, 'es'],
 [88514, 'ko'],
 [88536, 'ro'],
 [88574, 'it'],
 [88587, 'ko'],
 [88604, 'es'],
 [88605, 'es'],
 [88637, 'id'],
 [88649, 'es'],
 [88672, 'pt'],
 [88679, 'es'],
 [88759, 'pt'],
 [88762, 'nl'],
 [88791, 'es'],
 [88803, 'uk'],
 [88808, 'ro'],
 [88882, 'fr'],
 [88920, 'fr'],
 [88946, 'es'],
 [88964, 'ro'],
 [89036, 'it'],
 [89059, 'es'],
 [89119, 'es'],
 [89127, 'es'],
 [89181, 'ro'],
 [89185, 'ro'],
 [89259, 'fr'],
 [89272, 'fr'],
 [89301, 'ro'],
 [89379, 'ro'],
 [89403, 'pl'],
 [89447, 'ro'],
 [89451, 'de'],
 [89538, 'fr'],
 [89560, 'zh-cn'],
 [89631, 'de'],
 [89675, 'sl'],
 [89683, 'fr'],
 [89689, 'tl'],
 [89695, 'es'],
 [89719, 'es'],
 [89723, 'fr'],
 [89731, 'fr'],
 [89734, 'so'],
 [89746, 'it'],
 [89752, 'es'],
 [89760, 'no'],
 [89771, 'ro'],
 [89808, 'ca'],
 [89830, 'af'],
 [89841, 'so'],
 [89860, 'fr'],
 [89861, 'fr'],
 [89904, 'ko'],
 [89947, 'ro'],
 [899

In [386]:
len(not_eng4)

1181

In [13]:
# try to get it all translated and updated.   PASS #4
# ---------------------------------------------------

# for i in not_eng4:
    
#     rec_num = i[0]
    
#     cmt = revP['comments'].iloc[rec_num]
#     cmt = cmt.replace(".", ",")
    
#     tr = translator.translate(cmt)
    
#     revP['comments'].iloc[rec_num] = tr.text
#     print(i)

In [447]:
# All the comments are translated!  So save this file so I don't lose them:
# -------------------------------------------------------------------------

# revP.to_csv('reviews_trans.csv')

In [16]:
# Reload the translated reviews file.
# -----------------------------------

revP_tr = pd.read_csv("reviews_trans.csv")


In [17]:
revP_tr.iloc[116931]

Unnamed: 0                                                  233771
id                                                     491828570.0
listing_id                                              21783333.0
date                                                       7/20/19
reviewer_id                                             58058922.0
reviewer_name                                              Giacomo
comments         Very spacious apartment, well equipped, furnis...
Name: 116931, dtype: object

In [55]:
revP_tr.sample(10)

Unnamed: 0           id  listing_id     date  reviewer_id  \
98207      215047  420116259.0  20457194.0   3/5/19  244109092.0   
77159      193999  472926816.0  18672130.0  6/20/19  107246652.0   
34425      151265  158426861.0  14760639.0   6/6/17   18681575.0   
62645      179485  423302110.0  17383973.0  3/13/19  245627178.0   
31787      148627  425998257.0  14556875.0  3/19/19  173707511.0   
50152      166992  202032926.0  15917248.0  10/9/17   75818946.0   
72838      189678  171439778.0  18377747.0  7/17/17  136208687.0   
40518      157358  138356123.0  15283809.0  3/19/17   42729067.0   
102882     219722  284607687.0  20760300.0   7/1/18  182860727.0   
115734     232574  278165165.0  21700748.0  6/17/18  190619800.0   

       reviewer_name                                           comments  
98207       Brittany  Was super excited to stay at this spot! Then I...  
77159          Jacob  Amazing stay. The space is beautiful and in a ...  
34425          Elize  To start with the good, this apartment is in a...  
62645          Nicki  Beautiful place, loved the garage parking. The...  
31787           Ryan  Beautiful place/atmosphere and view of Chicago...  
50152       Kathleen  This apartment was perfect! It was bigger than...  
72838           Joel  Great hosts with a lovely home in a perfect lo...  
40518            Ben  Great place, great location. Irena was very he...  
102882     Stephanie  Lulu is an outstanding host! Although she live...  
115734          Greg  We realy enjoyed our stay with Felicia, she ha...

In [56]:
revP_tr.head(10)

Unnamed: 0          id  listing_id     date reviewer_id reviewer_name  \
0     116840  69989174.0  12170773.0  4/15/16  65467524.0     Elizabeth   
1     116841  70277659.0  12170773.0  4/17/16  56990266.0       Pranith   
2     116842  70615667.0  12170773.0  4/19/16  66514184.0         Katie   
3     116843  71131787.0  12170773.0  4/23/16  46343079.0         Alana   
4     116844  71376925.0  12170773.0  4/25/16  61131477.0          Beth   
5     116845  71724931.0  12170773.0  4/28/16  66887648.0         Erica   
6     116846  72038360.0  12170773.0  4/30/16  65964518.0         Karis   
7     116847  73030532.0  12170773.0   5/6/16  69353432.0         Katie   
8     116848  73789411.0  12170773.0  5/10/16  61768450.0       Heather   
9     116849  73884491.0  12170773.0  5/11/16  66047309.0        Johnna   

                                            comments  
0  The condo was amazing. It was very spacious an...  
1  Jessica was available on texts most of the tim...  
2  Jessica is great, fast response! Our place was...  
3  My family and I loved this apartment. It was v...  
4  We have a lovely weekend in this rental! It wa...  
5  The location and apartment were great, exactly...  
6  It's a luxury and stylish apartment. We had a ...  
7  We stayed three nights at the Archways at Linc...  
8  My husband and I stayed at the Archways at Lin...  
9  The location was perfect.  Apartment was welco...

LOOK AT THE OTHER FILES

In [434]:
host.iloc[1]

id                                                 5775
host_url         https://www.airbnb.com/users/show/5775
name                                   Craig & Kathleen
host_since                                     12/29/08
location               Chicago, Illinois, United States
response_time                        within a few hours
superhost                                             t
Name: 1, dtype: object

In [435]:
lst.iloc[1]

id                                                                          4505
listing_url                                    https://www.airbnb.com/rooms/4505
name                           One Great Apartment. 384 Great Reviews. 1 Bad ...
description                    We took the winter off, made things nicer arou...
host_id                                                                     5775
host_since                                                              12/29/08
neighbourhood                                                     Little Village
city                                                                     Chicago
state                                                                         IL
zipcode                                                                    60608
country_code                                                                  US
latitude                                                                41.85495
longitude                   

In [433]:
lst.loc[lst['id'] == 12170773]

id                            listing_url  \
1241  12170773  https://www.airbnb.com/rooms/12170773   

                                                   name  \
1241  Apartment in a Century Building Across from Li...   

                                            description   host_id host_since  \
1241  The remains of Art Deco arches lend character ...  10363033    12/1/13   

     neighbourhood     city state zipcode  ... cleaning_fee  \
1241      Old Town  Chicago    IL   60614  ...       $55.00   

      review_scores_rating  review_scores_accuracy review_scores_cleanliness  \
1241                  98.0                    10.0                      10.0   

     review_scores_checkin  review_scores_communication  \
1241                  10.0                         10.0   

      review_scores_location  review_scores_value instant_bookable  \
1241                    10.0                 10.0                f   

              cancellation_policy  
1241  strict_14_with_grace_period  

[1 rows x 29 columns]

In [436]:
avail.iloc[1]

listing_id            2384
date              11/22/19
available                f
price              $65.00 
minimum_nights           2
maximum_nights        1125
Name: 1, dtype: object

In [437]:
revP.iloc[1]

id                                                      70277659.0
listing_id                                              12170773.0
date                                                       4/17/16
reviewer_id                                             56990266.0
reviewer_name                                              Pranith
comments         Jessica was available on texts most of the tim...
Name: 116841, dtype: object

Here's how to link files.

In [438]:
revP.loc[revP['listing_id'] == 12170773]

id  listing_id      date  reviewer_id reviewer_name  \
116840   69989174.0  12170773.0   4/15/16   65467524.0     Elizabeth   
116841   70277659.0  12170773.0   4/17/16   56990266.0       Pranith   
116842   70615667.0  12170773.0   4/19/16   66514184.0         Katie   
116843   71131787.0  12170773.0   4/23/16   46343079.0         Alana   
116844   71376925.0  12170773.0   4/25/16   61131477.0          Beth   
...             ...         ...       ...          ...           ...   
117161  560088168.0  12170773.0   11/5/19  133068943.0         Julie   
117162  560383496.0  12170773.0   11/6/19   53715907.0         Megan   
117163  561482376.0  12170773.0   11/9/19   27761453.0        Dwight   
117164  562328014.0  12170773.0  11/10/19  112564219.0       Shannon   
117165  563340073.0  12170773.0  11/12/19  144873751.0       Lindsey   

                                                 comments  
116840  The condo was amazing. It was very spacious an...  
116841  Jessica was available on texts most of the tim...  
116842  Jessica is great, fast response! Our place was...  
116843  My family and I loved this apartment. It was v...  
116844  We have a lovely weekend in this rental! It wa...  
...                                                   ...  
117161  My sister and I took our daughters to Chicago ...  
117162  Excellent location and communicative & accommo...  
117163                                Great place to stay  
117164  Jessica made booking extremely easy and was ex...  
117165  Loved staying at Jessica’s place! It’s a great...  

[326 rows x 6 columns]

In [57]:
revP_tr.tail(10)

Unnamed: 0  \
117059                                             233899   
117060                                             233900   
117061                                             233901   
117062                                             233902   
117063                                             233903   
117064                                               1281   
117065                                                 42   
117066                                                 64   
117067                                               1940   
117068  房东是很和善的一对夫妻｡他们有很好的沟通能力,在我启程之前,他们用了一些时间来指导我顺利到达...   

                                                       id  \
117059                                        556868477.0   
117060                                        559780590.0   
117061                                        566148318.0   
117062                                        215412883.0   
117063                                        217997495.0   
117064                                                NaN   
117065                                                NaN   
117066                                                NaN   
117067                                                NaN   
117068  The landlord is a very kind couple. They have ...   

                                               listing_id  \
117059                                         21783531.0   
117060                                         21783531.0   
117061                                         21783531.0   
117062                                         21784716.0   
117063                                         21784716.0   
117064                                                NaN   
117065                                                NaN   
117066                                                NaN   
117067                                                NaN   
117068  The landlord is a very kind couple. They have ...   

                                                     date  \
117059                                           10/30/19   
117060                                            11/4/19   
117061                                           11/18/19   
117062                                           11/29/17   
117063                                           12/10/17   
117064                                                NaN   
117065                                                NaN   
117066                                                NaN   
117067                                                NaN   
117068  The landlord is a very kind couple. They have ...   

                                              reviewer_id  \
117059                                          8520589.0   
117060                                         72254229.0   
117061                                         42144708.0   
117062                                         13109574.0   
117063                                         46340737.0   
117064                                                NaN   
117065                                                NaN   
117066                                                NaN   
117067                                                NaN   
117068  The landlord is a very kind couple. They have ...   

                                            reviewer_name  \
117059                                            Lindsey   
117060                                             Howard   
117061                                             Rafael   
117062                                             Justin   
117063                                                Zac   
117064                                                NaN   
117065                                                NaN   
117066                                                NaN   
117067                                                NaN   
117068  The landlord is a very kind couple. They have ...   

               

Something I did resulted in some junk records at the end of the file.  Last record should be 117063.  Drop them.

In [61]:
# revP_tr.drop(revP_tr.tail(5).index, inplace=True)
revP_tr.tail(10)

Unnamed: 0           id  listing_id      date  reviewer_id  \
117054     233894  528783185.0  21783531.0   9/13/19   29594925.0   
117055     233895  535930162.0  21783531.0   9/25/19   18359595.0   
117056     233896  553250097.0  21783531.0  10/23/19  269173257.0   
117057     233897  554042083.0  21783531.0  10/25/19   18226663.0   
117058     233898  556139248.0  21783531.0  10/28/19   61956933.0   
117059     233899  556868477.0  21783531.0  10/30/19    8520589.0   
117060     233900  559780590.0  21783531.0   11/4/19   72254229.0   
117061     233901  566148318.0  21783531.0  11/18/19   42144708.0   
117062     233902  215412883.0  21784716.0  11/29/17   13109574.0   
117063     233903  217997495.0  21784716.0  12/10/17   46340737.0   

       reviewer_name                                           comments  
117054        Taylor  Great apartment for a work trip and in an amaz...  
117055       Elliott  Drew's place is the perfect location if you ar...  
117056     Christian  Nice modern apartment, very comfy and well equ...  
117057         Laura  Drew’s place is the perfect fit for visits to ...  
117058         Ravae  100% recommend. Lovely apartment and great loc...  
117059       Lindsey  Super clean listing that's very accessible to ...  
117060        Howard  Great Location! We will definitely book this u...  
117061        Rafael  Great location, quite space, has a king size b...  
117062        Justin  (Website hidden by Airbnb) where do I begin? L...  
117063           Zac  Awesome place, great neighborhood. It's right ...

OK.  Now start the NLP on the reviews.